### PR TITLE
Fix argmax, findmax, findXwithfirst, and expand testing

### LINKED ISSUE #101: findmax is broken for repeated values
@quinnj `findmax` is still broken when there are repeated values.

```
julia> c = ChainedVector([[7, 21, 31], [45], [53, 53]])
6-element ChainedVector{Int64, Vector{Int64}}:
  7
 21
 31
 45
 53
 53

julia> findmax(c)
(53, 6)

julia> findmax(c |> collect)
(53, 5)
```

Fixed by #99

### PR DESCRIPTION
**Pull request by mkitti**
This pull request adds more tests including randomly generated `ChainVector`s of `Int`, `Float64`, and `Rational`, incorporates #98 .

Functions fixed
* <strike>`argmin`</strike> (Already fixed)
* `argmax`
* <strike>`findmin`</strike> (Already fixed)
* `findmax`
* `findXwithfirst`

Fixed #101

### CODE REVIEW

File: test/chainedvector.jl [hunk 1]
```diff
@@ -423,6 +436,110 @@
...
+            @test findmin(x->x-1, x) == findmin(x->x-1, y)
+            @test findfirst(isodd, x) == findfirst(isodd, y)
+            @test findfirst(iseven, x) == findfirst(iseven ,y)
+            @test findlast(isodd, x) == findlast(isodd, y)
+            @test findlast(iseven, x) == findlast(iseven ,y)
+            @test findnext(isodd, x, 5) == findnext(isodd, y, 5)
+        catch err
+            error("Test failed for $x")
+        end
+    end
```
Comment (Seelengrab) [inline test/chainedvector.jl:539]:
```suggestion
    @testset for vecs in test_vectors
        x, y = vecs
        @test length(x) == length(y)
        # should this be approx?
        @test sum(x) ≈ sum(y)
        @test findmax(x) == findmax(y)
        @test findmin(x) == findmin(y)
        @test maximum(x) == maximum(y)
        @test minimum(x) == minimum(y)
        @test argmax(x) == argmax(y)
        @test argmin(x) == argmin(y)
        @test findmax(x->x+1, x) == findmax(x->x+1, y)
        @test findmin(x->x-1, x) == findmin(x->x-1, y)
        @test findfirst(isodd, x) == findfirst(isodd, y)
        @test findfirst(iseven, x) == findfirst(iseven ,y)
        @test findlast(isodd, x) == findlast(isodd, y)
        @test findlast(iseven, x) == findlast(iseven ,y)
        @test findnext(isodd, x, 5) == findnext(isodd, y, 5)
    end
```

This syntax should be supported already by Julia 1.3; might be good to replace the big double setup with just the `ChainedVector`s and `collect` them explicitly in the loop here.

Comment (mkitti) [reply test/chainedvector.jl:539]: The main reason I'm not using collect is because I don't trust it to be correct here.

Comment (mkitti) [reply test/chainedvector.jl]: Resolved in 
b5a923f

Comment (Seelengrab) [reply test/chainedvector.jl]: I think we can assume it to be correct

```
julia> using Supposition

julia> using SentinelArrays

julia> data_vector = Data.Vectors(Data.Integers{UInt8}(); max_size=100);

julia> vecs = Data.Vectors(data_vector; min_size=1, max_size=10);

julia> chainedvecs = map(vecs) do vs
           all(isempty, vs) && reject!()
           ChainedVector(vs)
       end;

julia> @check max_examples=100_000 collect_consistent(cv=chainedvecs) -> all(splat(==), zip(cv, collect(cv)));
Test Summary:      | Pass  Total  Time
collect_consistent |    1      1  0.9s
```

**Comment by Seelengrab** [comment]
Ah yes, the inconsistency of `argmax(x)` vs `argmax(foo, x)` strikes again... I'll have to get around to rebasing https://github.com/JuliaLang/julia/pull/41339

File: src/chainedvector.jl [hunk 2]
```diff
@@ -809,7 +809,7 @@ function Base.findmax(f::F, x::ChainedVector) where {F}
     cleanup!(x) # get rid of any empty arrays
     i = 1
     y = f(x.arrays[1][1])
-    return findXwithfirst(!isless, f, x, y, i)
+    return findXwithfirst((x,y)->isless(y,x), f, x, y, i)
```
Comment (nalimilan) [inline src/chainedvector.jl:812]: What's the point of this? At least deserves a comment.

File: test/chainedvector.jl [hunk 3]
```diff
@@ -423,6 +436,163 @@
...
+    for i in 1:10
+        test_vector_generator() |>
+            repr |>
+            x->replace(x, "=>" => "=>\n    ") |>
+            x->println(x,",")
+    end
+    =#
+
+    # Pairs of text vectors
```
Comment (nalimilan) [inline test/chainedvector.jl:468]: Why "text"?

File: test/chainedvector.jl [hunk 4]
```diff
@@ -423,6 +436,163 @@
...
+        ChainedVector(Vector{Rational{Int64}}[[9//10, 3//10, 1//10, 2//5], [4//5], [9//10, 2//5]]) =>
+            Rational{Int64}[9//10, 3//10, 1//10, 2//5, 4//5, 9//10, 2//5],
+
+    ]
+    @testset for (x,y) in Iterators.flatten([int_vectors, floating_point_vectors, rational_vectors])
+        @test copy(x) == y
+        @test collect(x) == y
+        @test length(x) == length(y)
+        # should this be approx?
+        @test sum(x) ≈ sum(y)
```
Comment (nalimilan) [inline test/chainedvector.jl:548]: I think `==` should work
```suggestion
        @test sum(x) == sum(y)
```

File: src/chainedvector.jl (hunk 2 above)
Comment (mkitti) [reply src/chainedvector.jl:812]: Would it be clearer if I wrote the following?

```
isgreater(x,y) = isless(y,x)
```

`!isless` is the wrong function. That is `>=`. We need `>`.

`x > y` is the same as `y < x`.

What would be helpful comment? Everything I just wrote perhaps?

File: test/chainedvector.jl (hunk 4 above)
Comment (mkitti) [reply test/chainedvector.jl:548]: It does not, that's why I added it.

File: src/chainedvector.jl (hunk 2 above)
Comment (mkitti) [inline src/chainedvector.jl:812]:
```suggestion
    # x > y iff y < x for a well ordered set
    # nb. isgreater = !isless is not correct. That is `>=`
    isgreater(x, y) = isless(y, x)
    return findXwithfirst(isgreater, f, x, y, i)
```

File: test/chainedvector.jl (hunk 4 above)
Comment (mkitti) [reply test/chainedvector.jl:548]: You know what, let's prove it. I'll commit it and see if CI passes.

File: test/chainedvector.jl [hunk 5]
```diff
@@ -423,6 +436,163 @@
...
+        test_vector_generator() |>
+            repr |>
+            x->replace(x, "=>" => "=>\n    ") |>
+            x->println(x,",")
+    end
+    =#
+
+    # Pairs of test vectors
+    # Some were inspired by https://github.com/JuliaData/SentinelArrays.jl/issues/97
+    int_vectors = [
```
Comment (nhz2) [inline test/chainedvector.jl:470]: Instead of a hard-coded set of test vectors, could a few 1000 be generated when the test runs? That might catch a few more edge cases.

Also, I think it would be clearer to only test `Int` here, to avoid `NaN` and floating point rounding.

**Comment by mkitti** [comment]
With `==` the failing tests in eee286e are

```
  (x, y) = ([-8.5, -1.2, -3.8, 7.5, 8.2, 7.5, -5.3, -2.7, 0.6, -6.2, 6.1, 1.4], [-8.5, -1.2, -3.8, 7.5, 8.2, 7.5, -5.3, -2.7, 0.6, -6.2, 6.1, 1.4])                                                                                                                                                           |   32     1     33  1.3s
  (x, y) = ([-7.2, 8.1, 2.3, 7.5, -8.4, -5.7], [-7.2, 8.1, 2.3, 7.5, -8.4, -5.7])                                                                                                                                                                                                                             |   32     1     33  0.0s
  (x, y) = ([-3.7, 7.8, -5.0, 0.1, 5.0, -4.1, -1.6, -0.9, 8.7, -7.8], [-3.7, 7.8, -5.0, 0.1, 5.0, -4.1, -1.6, -0.9, 8.7, -7.8])                                                                                                                                                                               |   32     1     33  0.0s
  (x, y) = ([8.6, -2.0, 8.0, 3.4, 3.3, 1.0, 5.4, -2.6, -4.7, 4.4, 4.4, 7.9], [8.6, -2.0, 8.0, 3.4, 3.3, 1.0, 5.4, -2.6, -4.7, 4.4, 4.4, 7.9])                                                                                                                                                                 |   32     1     33  0.0s
  (x, y) = ([7.6, 5.9, 7.9, -8.8, -1.5, -0.4, 6.0, -5.1, -0.4, 4.4, 7.3], [7.6, 5.9, 7.9, -8.8, -1.5, -0.4, 6.0, -5.1, -0.4, 4.4, 7.3])                                                                                                                                                                       |   32     1     33  0.0s
  (x, y) = ([3.2, -3.2, 1.2, -1.2, -2.1, 0.5, 6.2, 2.9, -8.1, 5.8, 4.8, -3.4, -3.1], [3.2, -3.2, 1.2, -1.2, -2.1, 0.5, 6.2, 2.9, -8.1, 5.8, 4.8, -3.4, -3.1])                                                                                                                                                 |   32     1     33  0.0s
  (x, y) = ([-7.5, -1.5, -5.8, 8.4, -8.4, -1.9, 2.3, -0.8, -8.5, 0.2, 0.5, -7.4, 2.1, -3.9], [-7.5, -1.5, -5.8, 8.4, -8.4, -1.9, 2.3, -0.8, -8.5, 0.2, 0.5, -7.4, 2.1, -3.9])                                                                                                                                 |   32     1     33  0.0s
```

File: test/chainedvector.jl (hunk 5 above)
Comment (mkitti) [reply test/chainedvector.jl:470]: I think that is part of @Seelengrab 's Supposition.jl is supposed to do. I'm just trying to get this pull request through before I start messing with adding test dependencies, but we seem to be stuck here.

File: test/chainedvector.jl [hunk 6]
```diff
@@ -423,6 +436,162 @@
...
+            Rational{Int64}[1//2, 1, 1//2, 9//10, 2//5, 9//10, 1//2, 3//5, 4//5, 7//10, 3//10, 2//5, 9//10, 1],
+        ChainedVector(Vector{Rational{Int64}}[[9//10, 3//10, 1//10, 2//5], [4//5], [9//10, 2//5]]) =>
+            Rational{Int64}[9//10, 3//10, 1//10, 2//5, 4//5, 9//10, 2//5],
+
+    ]
+    @testset for (x,y) in Iterators.flatten([int_vectors, floating_point_vectors, rational_vectors])
+        @test copy(x) == y
+        @test collect(x) == y
+        @test length(x) == length(y)
+        @test sum(x) == sum(y)
```
Comment (mkitti) [inline test/chainedvector.jl:547]:
```suggestion
        # Floating point tests fail if this is not approx
        # See https://github.com/JuliaData/SentinelArrays.jl/pull/99#issuecomment-2171005657
        @test sum(x) ≈ sum(y)
```

**Comment by mkitti** [comment]
Wait... why is sum failing?

```
y = ChainedVector([[-8.5, -1.2, -3.8, 7.5], [8.2, 7.5, -5.3], [-2.7, 0.6, -6.2, 6.1, 1.4]])
x = collect(y)

julia> sum(y) == sum(x)
false
```

Should `mapreduce` be defined like this?

https://github.com/JuliaData/SentinelArrays.jl/blob/d13560a5639f4ecfd07cd9fb39cd1bae5a0aecea/src/chainedvector.jl#L784

Why not just use `Iterators.flatten`?

```
julia> Base.mapreduce(f::F, op::OP, x::ChainedVector) where {F, OP} = mapreduce(f, op, Iterators.flatten(x.arrays))

julia> sum(y) == sum(x)
true
```

**Review by nhz2 — APPROVED** [review]
These changes look good. I think `sum` is allowed to perform the reduction in any order, so it's fine if it is only approximately equal.

**Comment by mkitti** [comment]
Currently thinking about

```julia
       function Base.mapreduce(f::F, op::OP, x::ChainedVector) where {F, OP}
           init = mapreduce(f, op, first(x.arrays))
           foreach(@view x.arrays[begin+1:end]) do a
               init = mapreduce(f, op, a; init)
           end
           return init
       end
```

This may be another pull request though.

**Review by quinnj — APPROVED** [review]
